ethcore/src/state/mod.rs

```diff
@@ -539,8 +542,49 @@ impl<B: Backend> State<B> {
 			|a| a.as_ref().and_then(|account| account.storage_root().cloned()))
 	}
 
-	/// Mutate storage of account `address` so that it is `value` for `key`.
-	pub fn storage_at(&self, address: &Address, key: &H256) -> TrieResult<H256> {
+	/// Get the value of storage at last checkpoint.
+	pub fn checkpoint_storage_at(&self, checkpoint_index: usize, address: &Address, key: &H256) -> TrieResult<Option<H256>> {
+		if let Some(ref checkpoint) = self.checkpoints.borrow().get(checkpoint_index) {
+			Ok(Some(if let Some(entry) = checkpoint.get(address) {
+				match entry {
+					Some(entry) => {
+						match entry.account {
+							Some(ref account) => {
+								if let Some(value) = account.cached_storage_at(key) {
```

Comment:
I don't get this, the function is only used in `reverted_storage_at` to initiate `origin` value of the storage key, so why are we looking for the new value value of the key (`cached_storage_at` query first the storage changes). Maybe it is `cached_original_storage_at` that was meant to be called?

Reply:
The checkpoint stores an account entry. When a reversion happens, it gets swapped with the current cache (or cause the current cache to get invalidated). So here, we first check whether we can get the storage key value in the checkpoint account entry, if not, then this value has not been changed, so we fetch the original value since last state commit.